storage: release replicated locks during intent resolution

nvb opened the pull request:
Fixes #109648.
Informs #100193.

This commit adds support for releasing replicated locks during point and ranged intent resolution, if any are found. Replicated locks are released if the resolving transaction is finalized or if resolving transaction is pending and the lock has been rolled back through an epoch bump or savepoint rollback.

Release note: None

nvb committed:
storage: release replicated locks during intent resolution

Fixes #109648.
Informs #100193.

This commit adds support for releasing replicated locks during point and ranged
intent resolution, if any are found. Replicated locks are released if the
resolving transaction is finalized or if resolving transaction is pending and
the lock has been rolled back through an epoch bump or savepoint rollback.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/110480)
<!-- Reviewable:end -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 9 of 9 files at r1, 1 of 1 files at r2, 5 of 5 files at r3, 2 of 2 files at r4, 7 of 7 files at r5, 1 of 1 files at r6, 3 of 3 files at r7, 5 of 5 files at r8, 9 of 9 files at r9, 4 of 4 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110480#-:-NecqlNQDJnE_viIuEX7:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten and @sumeerbhola)

___
*[`pkg/storage/mvcc.go` line 5389 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/110480#-NecpWZd-Ablybvvh-dp:-NecpWZd-Ablybvvh-dq:b-jhxb3f) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a3652f8ab7995d055513fe770e6862440bee7ee/pkg/storage/mvcc.go#L5389)):*
> ```Go
> 	writer Writer,
> 	ms *enginepb.MVCCStats,
> 	intent roachpb.LockUpdate,
> ```

nit: consider renaming this to something other than intent in this patch itself. Maybe the "resolution", from a similar TODO you had in the previous PR? 
___
*[`pkg/storage/testdata/mvcc_histories/replicated_locks` line 559 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/110480#-NecqbvS6u353nW1FLnn:-NecqbvS6u353nW1FLno:br9fdup) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a3652f8ab7995d055513fe770e6862440bee7ee/pkg/storage/testdata/mvcc_histories/replicated_locks#L559)):*
> ```
> lock (Replicated): "k3"/Exclusive -> txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=1 ts=10.000000000,0 min=0,0 seq=1} ts=0,0 del=false klen=0 vlen=0 mergeTs=<nil> txnDidNotUpdateMeta=false
> 
> # Pending resolution of the lock in the new epoch without a savepoint rollback
> ```

Should we add a test for resolution from a prior epoch being a no-op as well?


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
change to mvcc.go looks good to me. I didn't look at the tests since @arulajmani is also reviewing.
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 4 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110480#-:-Nee9JD16VIK5e789AKT:b-on20x3)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

nvb committed:
storage: add intent resolution results to TestMVCCHistories output

This commit adds the results of resolve_intent and resolve_intent_range
to the TestMVCCHistories output. This includes the number of keys resolved
and resume span information, if any.

Epic: None
Release note: None

nvb reviewed (commented):
TFTRs!

I updated the PR to address the TODO about lock pagination across multiple locks on the same locked key, ensuring that all locks on a key are resolved before returning a resume span. To ensure this was tested sufficiently, I added a prep commit which adds intent resolution results (num keys, num bytes, resume span) to `TestMVCCHistories` output.

This PR has two LGTMs, but I'll hold off on merging until @arulajmani is able to take one last look at these changes.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110480#-:-Nf2pu8640cw4PSVOQ_1:b-8ziaaw)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @arulajmani and @sumeerbhola)

___
*[`pkg/storage/mvcc.go` line 5389 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/110480#-NecpWZd-Ablybvvh-dp:-Nf2Wq43C_yu7sKA6jLX:b-p7h1gt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a3652f8ab7995d055513fe770e6862440bee7ee/pkg/storage/mvcc.go#L5389)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: consider renaming this to something other than intent in this patch itself. Maybe the "resolution", from a similar TODO you had in the previous PR?
</blockquote></details>

Done. I ended up going with "update", because that's the naming we use in `cmd_resolve_intent.go` and `cmd_resolve_intent_range.go`.
___
*[`pkg/storage/testdata/mvcc_histories/replicated_locks` line 559 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/110480#-NecqbvS6u353nW1FLnn:-Nf2Yean0-LGN5-hDtiF:b-v0r75y) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a3652f8ab7995d055513fe770e6862440bee7ee/pkg/storage/testdata/mvcc_histories/replicated_locks#L559)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we add a test for resolution from a prior epoch being a no-op as well?
</blockquote></details>

Done. This also revealed a bug where resolution from a prior epoch with ignored sequence numbers was able to release a lock from a newer epoch. Fixed and added a test for this as well.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 52 of 52 files at r11, 30 of 30 files at r12, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110480#-:-NfBy1vu0D0ezevCBiJg:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[`pkg/storage/testdata/mvcc_histories/replicated_locks` line 559 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/110480#-NecqbvS6u353nW1FLnn:-NfBxr3H0byvksRACQ4g:b-54k27y) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a3652f8ab7995d055513fe770e6862440bee7ee/pkg/storage/testdata/mvcc_histories/replicated_locks#L559)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Done. This also revealed a bug where resolution from a prior epoch with ignored sequence numbers was able to release a lock from a newer epoch. Fixed and added a test for this as well.
</blockquote></details>

Nice! I think this is the same one we saw together way back in April, right?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/110480#-:-NfByDnN2nSpXeIDzZcI:b-oduykh)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani)

___
*[`pkg/storage/testdata/mvcc_histories/replicated_locks` line 559 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/110480#-NecqbvS6u353nW1FLnn:-NfByFTq6kAkQGz2qGoe:b31anpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a3652f8ab7995d055513fe770e6862440bee7ee/pkg/storage/testdata/mvcc_histories/replicated_locks#L559)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Nice! I think this is the same one we saw together way back in April, right?
</blockquote></details>

Yep, I think so! It's an easy mistake to make.


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
```
docker: received unexpected HTTP status: 503 Service Unavailable.
```

bors r+
